File: assignments/objective-c/word-count/WordCountTest.m
#import <XCTest/XCTest.h>
#import "WordCount.h"

@interface test_suite : XCTestCase

@end

@implementation test_suite

- (void)testCountOneWord {
  WordCount *words = [[WordCount alloc] initWithString:@"word"];
  NSDictionary *expected = @{ @"word" : @1 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"word"],[result objectForKey:@"word"]);
}

- (void)testCountOneOfEeach {
  WordCount *words = [[WordCount alloc] initWithString:@"one of each"];
  NSDictionary *expected = @{ @"one" : @1, @"of" : @1, @"each" : @1 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"one"],[result objectForKey:@"one"]);
  XCTAssertEqual([expected objectForKey:@"of"],[result objectForKey:@"of"]);
  XCTAssertEqual([expected objectForKey:@"each"],[result objectForKey:@"each"]);
}

- (void)testCountMultipleOccurrences {
  WordCount *words = [[WordCount alloc] initWithString:@"one fish two fish red fish blue fish"];
  NSDictionary *expected = @{ @"one" : @1, @"fish" : @4, @"two" : @1, @"red" : @1, @"blue" : @1 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"one"],[result objectForKey:@"one"]);
  XCTAssertEqual([expected objectForKey:@"fish"],[result objectForKey:@"fish"]);
  XCTAssertEqual([expected objectForKey:@"two"],[result objectForKey:@"two"]);
  XCTAssertEqual([expected objectForKey:@"red"],[result objectForKey:@"red"]);
  XCTAssertEqual([expected objectForKey:@"blue"],[result objectForKey:@"blue"]);
}

- (void)testIgnorePunctation {
  WordCount *words = [[WordCount alloc] initWithString:@"car : carpet as java : javascript!!&@$%^&"];
  NSDictionary *expected = @{ @"car" : @1, @"carpet" : @1, @"as" : @1, @"java" : @1, @"javascript" : @1 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"car"],[result objectForKey:@"car"]);
  XCTAssertEqual([expected objectForKey:@"carpet"],[result objectForKey:@"carpet"]);
  XCTAssertEqual([expected objectForKey:@"as"],[result objectForKey:@"as"]);
  XCTAssertEqual([expected objectForKey:@"java"],[result objectForKey:@"java"]);
  XCTAssertEqual([expected objectForKey:@"javascript"],[result objectForKey:@"javascript"]);
}

- (void)includeNumbers {
  WordCount *words = [[WordCount alloc] initWithString:@"testing, 1, 2 testing"];
  NSDictionary *expected = @{ @"testing" : @2, @"1" : @1, @"2" : @1 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"testing"],[result objectForKey:@"testing"]);
  XCTAssertEqual([expected objectForKey:@"1"],[result objectForKey:@"1"]);
  XCTAssertEqual([expected objectForKey:@"2"],[result objectForKey:@"2"]);
}

- (void)normalizeCase {
  WordCount *words = [[WordCount alloc] initWithString:@"go Go GO"];
  NSDictionary *expected = @{ @"go" : @3 };
  NSDictionary *result = [words count];

  XCTAssertEqual([expected objectForKey:@"go"],[result objectForKey:@"go"]);
}

@end

File: assignments/java/word-count/src/test/java/WordCountTest.java
import org.junit.Test;

import java.lang.Integer;
import java.lang.String;
import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.*;

public class WordCountTest {

    private final WordCount wordCount = new WordCount();

    @Test
    public void countOneWord() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("word", 1);

        actualWordCount = wordCount.Phrase("word");
        assertEquals(
            expectedWordCount, actualWordCount
        );
    }

    @Test
    public void countOneOfEach() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("one", 1);
        expectedWordCount.put("of", 1);
        expectedWordCount.put("each", 1);

        actualWordCount = wordCount.Phrase("one of each");
        assertEquals(
            expectedWordCount, actualWordCount
        );
    }

    @Test
    public void countMultipleOccurences() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("one", 1);
        expectedWordCount.put("fish", 4);
        expectedWordCount.put("two", 1);
        expectedWordCount.put("red", 1);
        expectedWordCount.put("blue", 1);

        actualWordCount = wordCount.Phrase("one fish two fish red fish blue fish");
        assertEquals(
            expectedWordCount, actualWordCount
        );
    }

    @Test
    public void ignorePunctuation() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("car", 1);
        expectedWordCount.put("carpet", 1);
        expectedWordCount.put("as", 1);
        expectedWordCount.put("java", 1);
        expectedWordCount.put("javascript", 1);

        actualWordCount = wordCount.Phrase("car : carpet as java : javascript!!&@$%^&");
        assertEquals(
            expectedWordCount, actualWordCount
        );

    }

    @Test
    public void includeNumbers() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("testing", 2);
        expectedWordCount.put("1", 1);
        expectedWordCount.put("2", 1);

        actualWordCount = wordCount.Phrase("testing, 1, 2 testing");
        assertEquals(
            expectedWordCount, actualWordCount
        );
    }

    @Test
    public void normalizeCase() {
        Map<String, Integer> actualWordCount = new HashMap<String, Integer>();
        final Map<String, Integer> expectedWordCount = new HashMap<String, Integer>();
        expectedWordCount.put("go", 3);

        actualWordCount = wordCount.Phrase("go Go GO");
        assertEquals(
            expectedWordCount, actualWordCount
        );
    }

}
